File: 2019/cpp/day_06b.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

std::tuple<bool, int> travel(
    const int steps, const std::string& cur, const std::string& dest,
    std::unordered_map<std::string, std::unordered_set<std::string>>&
        is_orbitted_by,
    std::unordered_map<std::string, std::string>& orbits,
    std::unordered_set<std::string>& reached_before) {
  if (cur == dest) {
    return {true, steps};
  }
  if (reached_before.find(cur) != reached_before.end()) {
    return {false, -1};
  }
  reached_before.insert(cur);
  if (is_orbitted_by.find(cur) != is_orbitted_by.end()) {
    for (const auto orbitting_object : is_orbitted_by[cur]) {
      const auto [reached, final_step_count] =
          travel(steps + 1, orbitting_object, dest, is_orbitted_by, orbits,
                 reached_before);
      if (reached) {
        return {reached, final_step_count};
      }
    }
  }
  if (orbits.find(cur) != orbits.end()) {
    const auto [reached, final_step_count] = travel(
        steps + 1, orbits[cur], dest, is_orbitted_by, orbits, reached_before);
    if (reached) {
      return {reached, final_step_count};
    }
  }
  return {false, -1};
}

int main(int argc, char* argv[]) {
  // Get input
  std::string input = "../input/day_06_input";
  if (argc > 1) {
    input = argv[1];
  }
  std::ifstream file(input);
  std::unordered_map<std::string, std::string> orbits;
  std::unordered_map<std::string, std::unordered_set<std::string>>
      is_orbitted_by;
  std::unordered_map<std::string, int> n_orbits;
  std::unordered_set<std::string> reached_before;

  // Solve
  std::string line;
  const std::string delim = ")";
  while (std::getline(file, line)) {
    const size_t delim_pos = line.find(delim);
    const std::string primary_object = line.substr(0, delim_pos);
    const std::string orbitting_object = line.substr(
        delim_pos + delim.size(), line.size() - delim_pos - delim.size());
    orbits[orbitting_object] = primary_object;
    if (is_orbitted_by.find(primary_object) == std::end(is_orbitted_by)) {
      is_orbitted_by.insert({primary_object, {}});
    }
    is_orbitted_by[primary_object].insert(orbitting_object);
  }

  const auto [reached, n_steps] =
      travel(0, "YOU", "SAN", is_orbitted_by, orbits, reached_before);
  std::cout << n_steps - 2 << '\n';
  return n_steps - 2;
}

File: 2019/python/day_06b.py
#!/usr/bin/python3

import sys


def travel(steps, cur, dest, is_orbitted_by, orbits, reached_before):
    if cur == dest:
        return True, steps
    if cur in reached_before:
        return False, -1
    reached_before.add(cur)
    if cur in is_orbitted_by.keys():
        for orbitting_object in is_orbitted_by[cur]:
            reached, final_step_count = travel(
                steps + 1,
                orbitting_object,
                dest,
                is_orbitted_by,
                orbits,
                reached_before,
            )
            if reached:
                return reached, final_step_count
    if cur in orbits.keys():
        reached, final_step_count = travel(
            steps + 1, orbits[cur], dest, is_orbitted_by, orbits, reached_before
        )
        if reached:
            return reached, final_step_count
    return False, -1


def main():
    # Get input
    file_name = "../input/day_06_input"
    if len(sys.argv) > 1:
        file_name = sys.argv[1]
    file = open(file_name)
    orbits = dict()
    is_orbitted_by = dict()
    n_orbits = dict()
    input = list()
    for line in file:
        line = line.strip("\n")
        input.append(line)

    # Solve
    for line in input:
        orbit = line.split(")")
        orbits[orbit[1]] = orbit[0]
        if orbit[0] not in is_orbitted_by.keys():
            is_orbitted_by[orbit[0]] = set()
        is_orbitted_by[orbit[0]].add(orbit[1])
    reached_before = set()
    reached, n_steps = travel(0, "YOU", "SAN", is_orbitted_by, orbits, reached_before)
    print(n_steps - 2)
    return n_steps - 2


if __name__ == "__main__":
    main()
